matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -297,33 +420,75 @@ private void computeTransitionProbabilities(TimeStep<GPXExtension, GPXEntry, Pat
-                final Path path = algo.calcPath(from.getQueryResult().getClosestNode(),
-                        to.getQueryResult().getClosestNode());
+
+                final Path path = algo.calcPath(from.queryResult.getClosestNode(),
+                        to.queryResult.getClosestNode());
+
                 if (path.isFound()) {
                     timeStep.addRoadPath(from, to, path);
+
+                    // The router considers unfavored virtual edges using edge penalties.
+                    // However, if a virtual edge is included in the computed path,
+                    // the path distance does not include any penalty distance. Hence, we
+                    // need to add our own penalty distance for every unfavored edge in the path.
+                    final double penalizedPathDistance = penalizedPathDistance(path,
+                            queryGraph.getUnfavoredVirtualEdges());
+
+                    logger.info("Path from: {}, to: {}, penalized path length: {}",
+                            from, to, penalizedPathDistance);
+
                     final double transitionLogProbability = probabilities
-                            .transitionLogProbability(path.getDistance(), linearDistance, timeDiff);
+                            .transitionLogProbability(penalizedPathDistance, linearDistance);
                     timeStep.addTransitionLogProbability(from, to, transitionLogProbability);
+                } else {
+                    logger.info("No path found for from: {}, to: {}", from, to);
                 }
+                queryGraph.clearUnfavoredStatus();
+
             }
         }
     }
 
-    private MatchResult computeMatchResult(List<SequenceState<GPXExtension, GPXEntry, Path>> seq,
-                                           List<GPXEntry> gpxList, List<QueryResult> allCandidates,
-                                           EdgeExplorer explorer) {
-        // every virtual edge maps to its real edge where the orientation is already correct!
-        // TODO use traversal key instead of string!
-        final Map<String, EdgeIteratorState> virtualEdgesMap = new HashMap<>();
-        for (QueryResult candidate : allCandidates) {
-            fillVirtualEdges(virtualEdgesMap, explorer, candidate);
+    /**
+     * Returns the path length plus a penalty for each used unfavored edge.
+     */
+    private double penalizedPathDistance(Path path,
+                                         List<VirtualEdgeIteratorState> penalizedVirtualEdges) {
+        final Set<Integer> penalizedEdges = new HashSet<>();
+        for (VirtualEdgeIteratorState edge : penalizedVirtualEdges) {
+            penalizedEdges.add(edge.getEdge());
         }
+        double totalPenalty = 0;
+        for (EdgeIteratorState edge : path.calcEdges()) {
+            if (penalizedEdges.contains(edge.getEdge())) {
+                totalPenalty += VIRTUAL_NODE_U_TURN_PENALTY;
+            }
+        }
```

Comment:
> wondering if a HashSet is overkill (and slow). At least we could use e.g. trove IntHashSet (on edge ID)

HashSet is super fast for smaller sets (in Ks or even millions)and as we do not put all nodes in there it should be preferred because an array requires to be of the size of the graph.

What we could slightly improve is doing `new HashSet(penalizedVirtualEdges.size())` ... but with [this change](https://github.com/graphhopper/graphhopper/pull/909/files#diff-e8229c3dacc8229f89953bce87a2805aR146) there is now a set so we do not need to copy it?